### DISCUSSION TITLE
Discussion about llama 3.2 vision

### DISCUSSION THREAD

**Discussion by ngxson** [danbev/learning-ai#8] [General]
Hi @danbev , thanks for the very detailed notes.

Re. your problem about n_vocab, what I observed is that the shape of tensors are:
- `language_model.model.embed_tokens.weight 	[128 264, 4 096] `
- `language_model.lm_head.weight 	         [128 256, 4 096] `

So as you found out, the output tensor has 8 less tokens than the embd tensor.

However, instead of modifying internal llama.cpp to handle this exception, I propose that upon converting safetensors to GGUF, we simply extend the lm_head to add these tokens. We can set these parameters to 0 (or `-inf`? I'm not sure yet) to make its logits always be small:

```py
# adding 8 rows to output tensor
added_rows = torch.full((8, 4096), float(0.0))
output_extended = torch.cat((output, added_rows), dim=0)
```

**Comment by ngxson** [comment]
An alternative solution to filling zero or -inf could be to duplicate the vector from the output row of `<|end_of_text|>` token.

The consequence would be that the output logits for `<|image|>` will be exactly the same with `<|end_of_text|>`

**Reply by danbev** [reply]
@ngxson Thank you for your comments/suggestions! 

I'll take a look at this soon. I actually ran into a different [issue](https://github.com/danbev/learning-ai/blob/main/notes/llama.cpp/llama-3-2-vision.md#new-vision-api-mllama-issue) when migrating the mllama code to the new vision api (it was a little premature to say it worked I'm afraid :disappointed:  ) . As soon has I have this working like the previous version I'll return to this above issue.

**Reply by ngxson** [reply]
Ok I'm having a look now

> I've verified that the pre-processing produces the same output

Already this is good, what I fear was that the preprocessing of mllama is not separated from the inference code.

Btw if you need help to compare output, I can provide you a cpp function to dump tensor data into file, then use python script to load it into `list[float]` then compare if they are closed to each other (within an epsilon for example)

> Things that are different are how the image patch embeddings are handled in the newst version. The actual embedding tensor are copied to the context like this

One trick is that that you can use `ggml_backend_tensor_get` to get the tensor into `float[]` array, then set it as `llama_batch` embedding like the old version. We need to try this, just to isolate the problem. Probably you run into the problem because `llama_batch` with tensor input is a bit buggy for now.

```cpp
auto * img_embd = llama_vision_get_output_tensor(ctx);
std::vector<float> output_debug(ggml_nelements(img_embd));
ggml_backend_tensor_get(img_embd, output_debug.data(), 0, ggml_nbytes(img_embd));
for (int row = 0; row < 10; row++) {
    int off = row * img_embd->ne[0];
    printf("... %f %f %f\n", output_debug[off], output_debug[off+1], output_debug[off+2]);
}
```

**Reply by ngxson** [reply]
> Probably you run into the problem because `llama_batch` with tensor input is a bit buggy for now.

A bit more clear in this, `ubatch` does not support "cutting" the tensor in half if it does not fit into the physical batch limit.

I'll need to work on this later.

**Reply by danbev** [reply]
Ah that is interesting and does sound like I might be running to this. I'll try what you suggested and set the image patch embedding on the `llama_batch` like the previous version does. Thanks!

**Reply by danbev** [reply]
I finally found the [issue](https://github.com/danbev/learning-ai/blob/main/notes/llama.cpp/llama-3-2-vision.md#user-content-wip) and this was my fault. 

So it might be the case that we can use `embd_tensor` for this after all,  and I'll give this a try next. I'll also rebase my branch and start looking at the vocab solution you mentioned above. And then I'll start going through this to clean up. I mainly wanted to something working and then be able to improve it, so there is a lot of clean up and improvements to be made.

**Reply by danbev** [reply]
I used your suggestion to add extend `lm_head` which seems to work nicely :+1: 

I did not get `embd_tensor` to work but I'll revisit this tomorrow and I ran out of time last week.

**Reply by ngxson** [reply]
I think for now you can temporary ignore the `embd_tensor` and just set the `float * batch.embd` as usual. Otherwise, you must set a high ubatch value so that the tensor don't get truncated.

Support "cutting" `embd_tensor` should be easy to add, just need to play with `ggml_view`. I'll have a look later on.
